web/src/filter.ts

```diff
@@ -726,6 +726,53 @@ export class Filter {
         return true;
     }
 
+    static adjusted_terms_if_moved(raw_terms: NarrowTerm[], message: Message): NarrowTerm[] | null {
+        if (message.type !== "stream") {
+            return null;
+        }
+
+        assert(typeof message.display_recipient === "string");
+        assert(typeof message.topic === "string");
+
+        const adjusted_terms = [];
+        let terms_changed = false;
+
+        for (const term of raw_terms) {
+            const adjusted_term = {...term};
+            if (
+                Filter.canonicalize_operator(term.operator) === "channel" &&
+                !util.lower_same(term.operand, message.display_recipient)
+            ) {
+                adjusted_term.operand = message.display_recipient;
+                terms_changed = true;
+            }
+
+            if (
+                Filter.canonicalize_operator(term.operator) === "topic" &&
+                !util.lower_same(term.operand, message.topic)
+            ) {
+                adjusted_term.operand = message.topic;
+                terms_changed = true;
+            }
+
+            adjusted_terms.push(adjusted_term);
+        }
+
+        if (!terms_changed) {
+            return null;
+        }
+
+        return adjusted_terms;
+    }
+
+    setup_filter(terms: NarrowTerm[]): void {
+        this._terms = this.fix_terms(terms);
+        this.is_adjusted_for_moved_with_target = !this.has_operator("with");
```

Comment:
Hmm, what's the definition of `is_adjusted_for_moved_with_target`? It looks like it just is whether there's an operator, not whether we've actually done an adjustment...

Reply:
Oh, I see, you're initializing it as "fully adjusted already" if no "with" is present here. Name feels a bit misleading.

Reply:
Renamed to `requires_adjustment_for_moved_with_target` with the opposite meaning, which doesn't have that problem.

Reply:
It also seems weird that this should happen in the second `setup_filter` call; we can move this line to the constructor. Did this set of cleanups:
``` diff
diff --git a/web/src/filter.ts b/web/src/filter.ts
index c4e2585d22..de757b42a5 100644
--- a/web/src/filter.ts
+++ b/web/src/filter.ts
@@ -294,11 +294,12 @@ export class Filter {
     _sorted_term_types?: string[] = undefined;
     _predicate?: (message: Message) => boolean;
     _can_mark_messages_read?: boolean;
-    is_adjusted_for_moved_with_target?: boolean;
+    requires_adjustment_for_moved_with_target?: boolean;
 
     constructor(terms: NarrowTerm[]) {
         this._terms = terms;
         this.setup_filter(terms);
+        this.requires_adjustment_for_moved_with_target = this.has_operator("with");
     }
 
     static canonicalize_operator(operator: string): string {
@@ -767,7 +768,6 @@ export class Filter {
 
     setup_filter(terms: NarrowTerm[]): void {
         this._terms = this.fix_terms(terms);
-        this.is_adjusted_for_moved_with_target = !this.has_operator("with");
         if (this.has_operator("channel")) {
             this._sub = stream_data.get_sub_by_name(this.operands("channel")[0]!);
         }
@@ -1329,11 +1329,12 @@ export class Filter {
     fix_terms(terms: NarrowTerm[]): NarrowTerm[] {
         terms = this._canonicalize_terms(terms);
         terms = this._fix_redundant_is_private(terms);
-        terms = this._fix_redundant_with_private(terms);
+        terms = this._fix_redundant_with_dm(terms);
         return terms;
     }
 
     _fix_redundant_is_private(terms: NarrowTerm[]): NarrowTerm[] {
+        // Every DM is a DM, so drop `is:dm` if on a DM conversation.
         if (!terms.some((term) => Filter.term_type(term) === "dm")) {
             return terms;
         }
@@ -1341,7 +1342,9 @@ export class Filter {
         return terms.filter((term) => Filter.term_type(term) !== "is-dm");
     }
 
-    _fix_redundant_with_private(terms: NarrowTerm[]): NarrowTerm[] {
+    _fix_redundant_with_dm(terms: NarrowTerm[]): NarrowTerm[] {
+        // Because DMs can't move, the `with` operator is a noop on a
+        // DM conversation.
         if (terms.some((term) => Filter.term_type(term) === "dm")) {
             return terms.filter((term) => Filter.term_type(term) !== "with");
         }
@@ -1490,7 +1493,7 @@ export class Filter {
     }
 
     try_adjusting_for_moved_with_target(message?: Message): void {
-        if (this.is_adjusted_for_moved_with_target) {
+        if (!this.requires_adjustment_for_moved_with_target) {
             return;
         }
 
@@ -1503,10 +1506,10 @@ export class Filter {
             return;
         }
 
-        const adjust_terms = Filter.adjusted_terms_if_moved(this._terms, message);
-        if (adjust_terms) {
+        const adjusted_terms = Filter.adjusted_terms_if_moved(this._terms, message);
+        if (adjusted_terms) {
             this.setup_filter(adjust_terms);
         }
-        this.is_adjusted_for_moved_with_target = true;
+        this.requires_adjustment_for_moved_with_target = false;
     }
 }
```